rclpy/rclpy/parameter_service.py

```diff
@@ -0,0 +1,117 @@
+        )
+        list_parameters_service_name = '/'.join((nodename, 'list_parameters'))
+        node.create_service(
+            ListParameters, list_parameters_service_name, self._list_parameters_callback
+        )
+        set_parameters_service_name = '/'.join((nodename, 'set_parameters'))
+        node.create_service(
+            SetParameters, set_parameters_service_name, self._set_parameters_callback
+        )
+        set_parameters_atomically_service_name = '/'.join((nodename, 'set_parameters_atomically'))
+        node.create_service(
+            SetParametersAtomically, set_parameters_atomically_service_name,
+            self._set_parameters_atomically_callback
+        )
+
+    def _describe_parameters_callback(self, request, response):
+        for name in request.names:
+            p = self._node.get_parameter(name)
+            response.descriptors.append(p.get_descriptor())
+        return response
+
+    def _get_parameters_callback(self, request, response):
+        for name in request.names:
+            p = self._node.get_parameter(name)
+            response.values.append(p.get_parameter_value())
+        return response
+
+    def _get_parameter_types_callback(self, request, response):
+        for name in request.names:
+            response.types.append(self._node.get_parameter(name).get_parameter_type())
+        return response
+
+    def _list_parameters_callback(self, request, response):
+        names_with_prefixes = []
+        for name in self._node._parameters.keys():
+            if '.' in name:
+                names_with_prefixes.append(name)
+                continue
+            elif request.prefixes:
+                continue
+            else:
+                response.result.names.append(name)
+        if 1 == request.depth:
+            return response
+
+        if not 0 == request.depth:
+            # A depth of zero indicates infinite depth
+            names_with_prefixes = filter(
+                lambda name: name.count('.') + 1 <= request.depth, names_with_prefixes
+            )
+        for name in names_with_prefixes:
+            if request.prefixes:
+                for prefix in request.prefixes:
+                    prefix_with_trailing_dot = '%s.' % (prefix)
+                    if name.startswith(prefix_with_trailing_dot):
+                        response.result.names.append(name)
+                        full_prefix = '.'.join(name.split('.')[0:-1])
+                        if full_prefix not in response.result.prefixes:
+                            response.result.prefixes.append(full_prefix)
+                        if prefix not in response.result.prefixes:
+                            response.result.prefixes.append(prefix)
+            else:
+                prefix = '.'.join(name.split('.')[0:-1])
+                if prefix not in response.result.prefixes:
+                    response.result.prefixes.append(prefix)
+                response.result.names.append(name)
```

Comment:
The definition of the result message is ambiguous IMO: https://github.com/ros2/rcl_interfaces/blob/f00fe22a36c8fdf6145d873173a4f9de0fdff131/rcl_interfaces/msg/ListParametersResult.msg.

It's unclear to me if the name of the "parameters under the prefix" should be:
- the full parameter names (of the parameters that have a prefix matching one of the requested ones)
- parameter names stripped of the prefix (of the parameters that have a prefix matching one of the requested ones)

The current implementation seems to match the implementation of rclcpp so I think it's good :+1:. But maybe it would be worth rephrasing or addressing the todo in the message definition. Or clarify what we expect the `prefixes` field in the response to be used for.

Reply:
maybe @wjwwood has more context regarding the use of the `prefixes` field in the result message?